File: java/src/main/java/it/giannibombelli/wsc2026/booking/infrastructure/SqliteBookingRepository.java
package it.giannibombelli.wsc2026.booking.infrastructure;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingStatus;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import org.jooq.DSLContext;
import org.jooq.Record;
import org.jooq.SQLDialect;
import org.jooq.impl.DSL;

import javax.sql.DataSource;
import java.util.Optional;
import java.util.UUID;

import static org.jooq.impl.DSL.field;
import static org.jooq.impl.DSL.table;



public final class SqliteBookingRepository implements BookingRepository {
    private static final org.jooq.Table<?> BOOKING = table("booking");
    private static final org.jooq.Field<String> ID = field("id", String.class);
    private static final org.jooq.Field<String> DESCRIPTION = field("description", String.class);
    private static final org.jooq.Field<String> GIFT_CARD_ID = field("gift_card_id", String.class);
    private static final org.jooq.Field<String> STATUS = field("status", String.class);

    private final DSLContext dsl;

    public SqliteBookingRepository(DataSource dataSource) {
        this.dsl = DSL.using(dataSource, SQLDialect.SQLITE);
    }

    @Override
    public void save(Booking booking) {
        String idStr = booking.id().value().toString();

        String statusStr = booking.status().name();

        int updated = dsl.update(BOOKING)
            .set(ID, idStr)
            .set(DESCRIPTION, booking.description().value())
            .set(GIFT_CARD_ID, booking.giftCardId().value().toString())
            .set(STATUS, statusStr)
            .where(ID.eq(idStr))
            .execute();

        if (updated == 0) {
            dsl.insertInto(BOOKING)
                .columns(ID, DESCRIPTION, GIFT_CARD_ID, STATUS)
                .values(idStr, booking.description().value(), booking.giftCardId().value().toString(), statusStr)
                .execute();
        }
    }

    @Override
    public Optional<Booking> findById(BookingId id) {
        String idStr = id.value().toString();

        var rows = dsl
            .select(ID, DESCRIPTION, GIFT_CARD_ID, STATUS)
            .from(BOOKING)
            .where(ID.eq(idStr))
            .fetch();

        if (rows.isEmpty()) {
            return Optional.empty();
        }

        Record r = rows.getFirst();
        BookingId bookingId = new BookingId(UUID.fromString(r.get(ID)));
        Description description = new Description(r.get(DESCRIPTION));
        GiftCardId giftCardId = new GiftCardId(UUID.fromString(r.get(GIFT_CARD_ID)));
        BookingStatus status = BookingStatus.valueOf(r.get(STATUS));
        return Optional.of(new Booking(bookingId, description, giftCardId, status));
    }
}

File: typescript/src/booking/infrastructure/sqliteBookingRepository.ts
// Adapter repository per SQLite usando Drizzle ORM.


import Database from 'better-sqlite3';
import { drizzle } from 'drizzle-orm/better-sqlite3';
import type { BetterSQLite3Database } from 'drizzle-orm/better-sqlite3';
import { eq } from 'drizzle-orm';

import { Description } from '@/common/domain/primitive/description.js';
import { GiftCardReference } from '../domain/primitive/giftCardReference.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { Booking } from '../domain/booking/booking.js';
import { BookingId } from '../domain/booking/bookingId.js';
import { BookingStatus } from '../domain/booking/bookingStatus.js';
import type { BookingRepository } from '../domain/ports/bookingRepository.js';
import { bookings } from './schema.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class SqliteBookingRepository implements BookingRepository {
  private readonly _db: BetterSQLite3Database;

  constructor(database: Database.Database) {
    requireDependency(database, "database");
    this._db = drizzle(database);
  }

  save(booking: Booking): void {
    const id = booking.id().value.value;
    const status = booking.status();

    this._db
      .insert(bookings)
      .values({
        id,
        description: booking.description().value,
        giftCardId: booking.giftCardReference().value.value,
        status,
      })
      .onConflictDoUpdate({
        target: bookings.id,
        set: {
          description: booking.description().value,
          giftCardId: booking.giftCardReference().value.value,
          status,
        },
      })
      .run();
  }

  findById(id: BookingId): Booking | null {
    const rows = this._db
      .select()
      .from(bookings)
      .where(eq(bookings.id, id.value.value))
      .all();

    if (rows.length === 0) {
      return null;
    }

    const row = rows[0];
    return new Booking(
      new BookingId(Uuid.fromString(row.id)),
      new Description(row.description),
      new GiftCardReference(Uuid.fromString(row.giftCardId)),
      BookingStatus[row.status as keyof typeof BookingStatus]
    );
  }
}
